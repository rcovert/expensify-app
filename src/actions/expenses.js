import uuid from "uuid";
import database from '../firebase/firebase';

// component calls action generator
// action generator returns object
// component dispatches object
// redux stores changes
//
// now change to
// component calls action generator
// action generatr returns function
// component dispatches function with redux middleware
// function runs - has the ability to dispath other actions etc...

//ADD_EXPENSE - action generator - function returns implicit object
// note use of destructoring operation - pulling items from object and creating variables
// export const addExpense = ({
//     description = '',
//     notes = '',
//     amount = 0,
//     createdAt = 0
// } = {})
//     => ({
//         type: 'ADD_EXPENSE',
//         expense: {
//             id: uuid(),
//             description,
//             notes,
//             amount,
//             createdAt
//         }
//     });
// simplified based on asynch function call below with thunk
// ADD_EXPENSE
// add asynch action -- passing function only works with thunk middleware
// note expenseData = {} sets {} as empty default
// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};


//REMOVE_EXPENSE
// set action generator = function that implicitly returns a function
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})
//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// atart set expenses - to get data from database
export const startSetExpenses = () => {
    return (dispatch) => {
        // retrieve the data from the database - note this return 
        // returns the promise that can then be used by caller - i.e. the then clause
        // in app.js for this call
        return database.ref('expenses')
            .once('value')
            .then((snapshot) => {
                //console.log(snapshot.val())
                const expenses = [];
                // snapshot for each returns child snapshot/indv objects
                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });
                // now set redux with the information
                dispatch(setExpenses(expenses));
            });
    };
};
