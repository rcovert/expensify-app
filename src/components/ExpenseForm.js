import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// old way of date manipulation const date = new Date();
// const now = moment();  // returns now
// console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        // set values to existing expense values for edit
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: undefined
        };
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description: description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        // updater function (() => ({object}))
        this.setState(() => ({ note }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        // validate amount
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        };
    };
    onDateChange = (createdAt) => {
        // check if there is a value - if not do nothing
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        // avoid full page refresh behavior 
        e.preventDefault();
        // perform validation on desc and amt
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please add a description and amount to the form.' }));
        } else {
            this.setState(() => ({ error: '' }));
            // now pass the valid data back to parent component for action dispatch
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };

    // react class components must have render method
    // note use of form as root element for page
    // does not have to be a div
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type='text'
                    className="text_input"
                    placeholder='Description'
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    type='text'
                    className="text_input"
                    placeholder='Amount'
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    placeholder='Add a note for your expense (optional)'
                    className="textarea"
                    value={this.state.note}
                    onChange={this.onNoteChange}>
                </textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>

            </form>
        )
    }
};
