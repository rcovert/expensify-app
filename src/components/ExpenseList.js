import React from 'react';
import { connect, Connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// note have to export this component for testing -- in addition to the
// default connected export below
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No Expense Items</p>
            ) : (
                props.expenses.map((expense) => (
                    <ExpenseListItem key={expense.id} {...expense} />
                ))
            )
        }
        
    </div>
)

const mapStateToProps = (state) => {
    // pass the expenses array as a prop/expenses to the expense list
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

// connect returns a function that then is called with ExpenseList to connect the
// component to the store
export default connect(mapStateToProps)(ExpenseList);
