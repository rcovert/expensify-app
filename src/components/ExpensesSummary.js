import React from 'react';
import { connect, Connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectTotalExpenses from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => (
    <div>
        {
            expenseCount === 0 ? (
                <p>No Expense Items</p>
            ) : (
                <p>There are {expenseCount} item(s), totaling:  
                  {numeral(expensesTotal/100).format(' $0,0.00')}</p>
            )
        }
    </div>
)

const mapStateToProps = (state) => {
    // pass the expenses array as a prop/expenses to the expense list
    // setup visibleExpenses as viewed array
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    // create object with new props to display in jsx above
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectTotalExpenses(visibleExpenses)
    };
};

// connect returns a function that then is called with ExpenseList to connect the
// component to the store
export default connect(mapStateToProps)(ExpensesSummary);