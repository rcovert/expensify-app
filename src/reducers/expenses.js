const expensesReducerDefaultState = [];
// Expenses reducer -- note use of ... spread operator to add to state
// note removal of const when defining this as the export default
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':  // use state.filter to get a new array back
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':  // use state.map to get a new array back
            // note individual items are expense
            return state.map((expense) => {
                if (expense.id === action.id) {
                    // note use of object spread and override syntax ...,...
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                };
            });
        default:
            return state;
    }
};

