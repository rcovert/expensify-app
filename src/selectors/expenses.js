import moment from 'moment';

// get visible expenses
// destructure filters to get individual variables
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'): true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day'): true;
        // convert text match source and target to lower case
        // and check includes for filter
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
        // note - sort is tacked onto return results from filter which is an array 
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            // if a is greater than b, then a comes first
            return a.amount < b.amount ? 1 : -1
        }
    })
}
