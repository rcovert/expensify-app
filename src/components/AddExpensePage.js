import React from "react";
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

// note props are passed in from connected object 

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // now dispath redux action console.log(expense);
        // now route back to the dashboard with history.push method from router/route
        // props.dispatch(addExpense(expense));
        // use this.props for class based components - not required on stateless functional components
        this.props.addExpense(expense);
        this.props.history.push('/');
    };
    render() {
        // note - this form of the code avoids writing function in-line
        // which prevents recalculation of function on every render
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    };
};

// stateless functional version
// const AddExpensePage = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//             onSubmit={(expense) => {
//                 props.onSubmit(expense);
//                 props.history.push('/');
//             }}
//         />
//     </div>
// );

// mapDispatchToProps is used to simplify code and make it easier for testing
const mapDispatchToProps = (dispatch) => ({
    // implicit return of object -- no need for return statement
    addExpense: (expense) => dispatch(addExpense(expense))
});
// mapStateToProps is first arg; mapDispatchToProps is 2nd arg
export default connect(undefined, mapDispatchToProps)(AddExpensePage);