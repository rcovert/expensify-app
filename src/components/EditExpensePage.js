import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense, startRemoveExpense } from '../actions/expenses';

// refactor to class based component
// setup mapDispatchToProps

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };
    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <h1>Edit Expense</h1>
                <p>The id is: {this.props.expense.id}</p>
                <ExpenseForm expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>
                    Remove</button>
            </div>
        )
    }
}

// const EditExpensePage = (props) => {
//     // console.log('the state at edit = ');
//     // console.log(props);
//     return (
//         <div>
//             <h1>Edit Expense</h1>
//             <p>The id is: {props.match.params.id}</p>
//             <ExpenseForm expense={props.expense}
//                 onSubmit={(expense) => {
//                     // now dispath redux action console.log(expense);
//                     // now route back to the dashboard with history.push method from router/route
//                     // console.log(expense.description);
//                     // console.log(props.match.params.id)
//                     props.dispatch(editExpense(props.match.params.id, expense));
//                     props.history.push('/');
//                 }}
//             />
//             <button onClick={() => {
//                 // console.log(props.match.params.id)
//                 // note that we are sending an object to remove expense
//                 props.dispatch(removeExpense({ id: props.match.params.id }))
//                 props.history.push('/');
//             }}>
//                 Remove</button>
//         </div>
//     )
// };

// this is where we need a map state to props function to get the expense with the right id
// which comes in on the match prop
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;
        })
    }
}

const mapDispatchToProps = (dispatch, props) => ({
    // implicit return of object -- no need for return statement
    // edit expense is set to a function
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    startRemoveExpense: (expense) => dispatch(startRemoveExpense(expense))
});
// need to pass in mapStateToProps to connect to get the props set
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);