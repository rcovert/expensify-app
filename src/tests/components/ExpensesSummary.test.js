import React from 'react';
import { shallow } from 'enzyme';
// be sure to use {} for named exports
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

// not sure why this did not work
// test('should render ExpenseSummary with expenses', () => {
//     const wrapper = shallow(<ExpensesSummary expenses={expenses} />);
//     expect(wrapper).toMatchSnapshot();
// });

// test('should render ExpenseSummary with no expeses', () => {
//     const wrapper = shallow(<ExpensesSummary expenses={[]} />);
//     expect(wrapper).toMatchSnapshot();
// });

// test('should render ExpenseSummary with just one expeses', () => {
//     const wrapper = shallow(<ExpensesSummary expenses={expenses[9]} />);
//     expect(wrapper).toMatchSnapshot();
// });

// note props match props in component map state to props
test('should correctly render ExpensesSummary with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={235} />);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={23} expensesTotal={23512340987} />);
    expect(wrapper).toMatchSnapshot();
});


