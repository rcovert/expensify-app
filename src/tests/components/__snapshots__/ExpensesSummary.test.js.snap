// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <p>
    There are 
    1
     item(s), totaling:
     $2.35
  </p>
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<div>
  <p>
    There are 
    23
     item(s), totaling:
     $235,123,409.87
  </p>
</div>
`;
